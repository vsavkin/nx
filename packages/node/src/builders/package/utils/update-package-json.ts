import { BuilderContext } from '@angular-devkit/architect';
import { normalize } from '@angular-devkit/core';
import { readJsonFile } from '@nrwl/workspace';
import { writeJsonFile } from '@nrwl/workspace/src/utilities/fileutils';
import { basename, join } from 'path';
import { NormalizedBuilderOptions } from './models';

export default function updatePackageJson(
  options: NormalizedBuilderOptions,
  context: BuilderContext
) {
  const mainFile = basename(options.main).replace(/\.[tj]s$/, '');
  const typingsFile = `${mainFile}.d.ts`;
  const mainJsFile = `${mainFile}.js`;
  const packageJson = readJsonFile(
    join(context.workspaceRoot, options.packageJson)
  );

  packageJson.main = normalize(
    `./${options.relativeMainFileOutput}/${mainJsFile}`
  );
  packageJson.typings = normalize(
    `./${options.relativeMainFileOutput}/${typingsFile}`
  );

  writeJsonFile(`${options.outputPath}/package.json`, packageJson);
}
