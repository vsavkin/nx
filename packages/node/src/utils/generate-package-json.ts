import { ProjectGraph, readJsonFile } from '@nrwl/workspace';
import { BuildNodeBuilderOptions } from './types';
import { writeJsonFile } from '@nrwl/workspace/src/utilities/fileutils';
import { OUT_FILENAME } from './config';

/**
 * Creates a package.json in the output directory for support  to install dependencies within containers.
 *
 * If a package.json exists in the project, it will reuse that.
 *
 * @param projectName
 * @param graph
 * @param options
 * @constructor
 */
export function generatePackageJson(
  projectName: string,
  graph: ProjectGraph,
  options: BuildNodeBuilderOptions
) {
  const npmDeps = findAllNpmDeps(projectName, graph);
  // default package.json if one does not exist
  let packageJson = {
    name: projectName,
    version: '0.0.1',
    main: OUT_FILENAME,
    dependencies: {},
  };

  try {
    packageJson = readJsonFile(`${options.projectRoot}/package.json`);
    if (!packageJson.dependencies) {
      packageJson.dependencies = {};
    }
  } catch (e) {}

  const rootPackageJson = readJsonFile(`${options.root}/package.json`);

  Object.entries(npmDeps).forEach(([packageName, version]) => {
    // don't include devDeps
    if (rootPackageJson.devDependencies?.[packageName]) {
      return;
    }

    packageJson.dependencies[packageName] = version;
  });

  writeJsonFile(`${options.outputPath}/package.json`, packageJson);
}

function findAllNpmDeps(
  projectName: string,
  graph: ProjectGraph,
  list: { [packageName: string]: string } = {},
  seen = new Set<string>()
) {
  if (seen.has(projectName)) {
    return list;
  }

  seen.add(projectName);

  const node = graph.nodes[projectName];

  if (node.type === 'npm') {
    list[node.data.packageName] = node.data.version;
  }
  graph.dependencies[projectName]?.forEach((dep) => {
    findAllNpmDeps(dep.target, graph, list, seen);
  });

  return list;
}
