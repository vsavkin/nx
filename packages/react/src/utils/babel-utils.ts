import { appRootPath } from '@nrwl/workspace/src/utilities/app-root';
import { readJsonFile } from '@nrwl/workspace';
import { join } from 'path';

export function updateBabelOptions(options: any): void {
  // Add react babel preset
  const idx = options.presets.findIndex(
    (p) => Array.isArray(p) && p[0].indexOf('@babel/preset-env') !== -1
  );
  options.presets.splice(idx + 1, 0, [
    require.resolve('@babel/preset-react'),
    {
      useBuiltIns: true,
    },
  ]);

  // TODO: Remove this once we have composable webpack and babel plugins.

  // Add babel plugin for styled-components or emotion.
  // We don't have a good way to know when a project uses one or the other, so
  // add the plugin only if the other style package isn't used.
  const packageJson = readJsonFile(join(appRootPath, 'package.json'));
  const deps = { ...packageJson.dependencies, ...packageJson.devDependencies };
  const hasStyledComponents = !!deps['styled-components'];
  const hasEmotion = !!deps['@emotion/react'];
  if (hasStyledComponents && !hasEmotion) {
    options.plugins.splice(0, 0, [
      require.resolve('babel-plugin-styled-components'),
      {
        pure: true,
      },
    ]);
  }
  if (hasEmotion && !hasStyledComponents) {
    options.plugins.splice(0, 0, require.resolve('babel-plugin-emotion'));
  }

  return options;
}
