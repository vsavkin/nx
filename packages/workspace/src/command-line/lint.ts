import {
  createProjectGraph,
  onlyWorkspaceProjects,
} from '../core/project-graph';
import { WorkspaceIntegrityChecks } from './workspace-integrity-checks';
import { readWorkspaceFiles, workspaceLayout } from '../core/file-utils';
import { output } from '../utilities/output';
import * as path from 'path';

export function workspaceLint() {
  const graph = onlyWorkspaceProjects(createProjectGraph());

  const cliErrorOutputConfigs = new WorkspaceIntegrityChecks(
    graph,
    readAllFilesFromAppsAndLibs()
  ).run();

  if (cliErrorOutputConfigs.length > 0) {
    cliErrorOutputConfigs.forEach((errorConfig) => {
      output.error(errorConfig);
    });
    process.exit(1);
  }
}

function readAllFilesFromAppsAndLibs() {
  const wl = workspaceLayout();
  return readWorkspaceFiles()
    .map((f) => f.file)
    .filter(
      (f) => f.startsWith(`${wl.appsDir}/`) || f.startsWith(`${wl.libsDir}/`)
    )
    .filter((f) => !path.basename(f).startsWith('.'));
}
