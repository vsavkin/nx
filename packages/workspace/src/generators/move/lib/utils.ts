import * as path from 'path';

import { ProjectConfiguration, Tree, getWorkspaceLayout } from '@nrwl/devkit';

import { Schema } from '../schema';

/**
 * This helper function ensures that we don't move libs or apps
 * outside of the folders they should be in.
 *
 * This will break if someone isn't using the default libs/apps
 * folders. In that case, they're on their own :/
 *
 * @param schema
 * @param workspace
 */
export function getDestination(
  host: Tree,
  schema: Schema,
  project: ProjectConfiguration
): string {
  const projectType = project.projectType;

  const workspaceLayout = getWorkspaceLayout(host);

  let rootFolder = workspaceLayout.libsDir;
  if (projectType === 'application') {
    rootFolder = workspaceLayout.appsDir;
  }
  return path.join(rootFolder, schema.destination);
}

/**
 * Replaces slashes with dashes
 *
 * @param path
 */
export function getNewProjectName(path: string): string {
  return path.replace(/\//g, '-');
}

/**
 * Normalizes slashes (removes duplicates)
 *
 * @param input
 */
export function normalizeSlashes(input: string): string {
  return input
    .split('/')
    .filter((x) => !!x)
    .join('/');
}
