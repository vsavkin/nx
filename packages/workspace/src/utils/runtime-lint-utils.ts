import * as path from 'path';
import { FileData } from '../core/file-utils';
import {
  DependencyType,
  isWorkspaceProject,
  ProjectGraph,
  ProjectGraphDependency,
  ProjectGraphNode,
} from '../core/project-graph';
import { TargetProjectLocator } from '../core/target-project-locator';

export type Deps = { [projectName: string]: ProjectGraphDependency[] };
export type DepConstraint = {
  sourceTag: string;
  onlyDependOnLibsWithTags: string[];
};

export function hasNoneOfTheseTags(proj: ProjectGraphNode, tags: string[]) {
  return tags.filter((allowedTag) => hasTag(proj, allowedTag)).length === 0;
}

function hasTag(proj: ProjectGraphNode, tag: string) {
  return (proj.data.tags || []).indexOf(tag) > -1 || tag === '*';
}

function containsFile(
  files: FileData[],
  targetFileWithoutExtension: string
): boolean {
  return !!files.filter(
    (f) => removeExt(f.file) === targetFileWithoutExtension
  )[0];
}

function removeExt(file: string): string {
  return file.replace(/\.[^/.]+$/, '');
}

function removeWindowsDriveLetter(osSpecificPath: string): string {
  return osSpecificPath.replace(/^[A-Z]:/, '');
}

function normalizePath(osSpecificPath: string): string {
  return removeWindowsDriveLetter(osSpecificPath).split(path.sep).join('/');
}

export function matchImportWithWildcard(
  // This may or may not contain wildcards ("*")
  allowableImport: string,
  extractedImport: string
): boolean {
  if (allowableImport.endsWith('/**')) {
    const prefix = allowableImport.substring(0, allowableImport.length - 2);
    return extractedImport.startsWith(prefix);
  } else if (allowableImport.endsWith('/*')) {
    const prefix = allowableImport.substring(0, allowableImport.length - 1);
    if (!extractedImport.startsWith(prefix)) return false;
    return extractedImport.substring(prefix.length).indexOf('/') === -1;
  } else if (allowableImport.indexOf('/**/') > -1) {
    const [prefix, suffix] = allowableImport.split('/**/');
    return (
      extractedImport.startsWith(prefix) && extractedImport.endsWith(suffix)
    );
  } else {
    return new RegExp(allowableImport).test(extractedImport);
  }
}

export function isRelative(s: string) {
  return s.startsWith('.');
}

export function isRelativeImportIntoAnotherProject(
  imp: string,
  projectPath: string,
  projectGraph: ProjectGraph,
  sourceFilePath: string
): boolean {
  if (!isRelative(imp)) return false;

  const targetFile = normalizePath(
    path.resolve(path.join(projectPath, path.dirname(sourceFilePath)), imp)
  ).substring(projectPath.length + 1);

  const sourceProject = findSourceProject(projectGraph, sourceFilePath);
  const targetProject = findTargetProject(projectGraph, targetFile);
  return sourceProject && targetProject && sourceProject !== targetProject;
}

export function findProjectUsingFile(projectGraph: ProjectGraph, file: string) {
  return Object.values(projectGraph.nodes).filter((n) =>
    containsFile(n.data.files, file)
  )[0];
}

export function findSourceProject(
  projectGraph: ProjectGraph,
  sourceFilePath: string
) {
  const targetFile = removeExt(sourceFilePath);
  return findProjectUsingFile(projectGraph, targetFile);
}

export function findTargetProject(
  projectGraph: ProjectGraph,
  targetFile: string
) {
  let targetProject = findProjectUsingFile(projectGraph, targetFile);
  if (!targetProject) {
    targetProject = findProjectUsingFile(
      projectGraph,
      normalizePath(path.join(targetFile, 'index'))
    );
  }
  if (!targetProject) {
    targetProject = findProjectUsingFile(
      projectGraph,
      normalizePath(path.join(targetFile, 'src', 'index'))
    );
  }
  return targetProject;
}

export function isAbsoluteImportIntoAnotherProject(imp: string) {
  // TODO: vsavkin: check if this needs to be fixed once we generalize lint rules
  return (
    imp.startsWith('libs/') ||
    imp.startsWith('/libs/') ||
    imp.startsWith('apps/') ||
    imp.startsWith('/apps/')
  );
}

export function findProjectUsingImport(
  projectGraph: ProjectGraph,
  targetProjectLocator: TargetProjectLocator,
  filePath: string,
  imp: string,
  npmScope: string
) {
  const target = targetProjectLocator.findProjectWithImport(
    imp,
    filePath,
    npmScope
  );
  return projectGraph.nodes[target];
}

export function checkCircularPath(
  graph: ProjectGraph,
  sourceProject: ProjectGraphNode,
  targetProject: ProjectGraphNode
): Array<ProjectGraphNode> {
  if (!graph.nodes[targetProject.name]) return [];
  return getPath(graph, targetProject.name, sourceProject.name);
}

interface Reach {
  graph: ProjectGraph;
  matrix: Record<string, Array<string>>;
  adjList: Record<string, Array<string>>;
}

const reach: Reach = {
  graph: null,
  matrix: null,
  adjList: null,
};

function buildMatrix(graph: ProjectGraph) {
  const dependencies = graph.dependencies;
  const nodes = Object.keys(graph.nodes).filter((s) =>
    isWorkspaceProject(graph.nodes[s])
  );
  const adjList = {};
  const matrix = {};

  const initMatrixValues = nodes.reduce((acc, value) => {
    return {
      ...acc,
      [value]: false,
    };
  }, {});

  nodes.forEach((v, i) => {
    adjList[nodes[i]] = [];
    matrix[nodes[i]] = { ...initMatrixValues };
  });

  for (let proj in dependencies) {
    for (let dep of dependencies[proj]) {
      if (isWorkspaceProject(graph.nodes[dep.target])) {
        adjList[proj].push(dep.target);
      }
    }
  }

  const traverse = (s, v) => {
    matrix[s][v] = true;

    for (let adj of adjList[v]) {
      if (matrix[s][adj] === false) {
        traverse(s, adj);
      }
    }
  };

  nodes.forEach((v, i) => {
    traverse(nodes[i], nodes[i]);
  });

  return {
    matrix,
    adjList,
  };
}

function getPath(
  graph: ProjectGraph,
  sourceProjectName: string,
  targetProjectName: string
): Array<ProjectGraphNode> {
  if (sourceProjectName === targetProjectName) return [];

  if (reach.graph !== graph) {
    const result = buildMatrix(graph);
    reach.graph = graph;
    reach.matrix = result.matrix;
    reach.adjList = result.adjList;
  }

  const adjList = reach.adjList;

  let path: string[] = [];
  const queue: Array<[string, string[]]> = [[sourceProjectName, path]];
  const visited: string[] = [sourceProjectName];

  while (queue.length > 0) {
    const [current, p] = queue.pop();
    path = [...p, current];

    if (current === targetProjectName) break;

    adjList[current]
      .filter((adj) => visited.indexOf(adj) === -1)
      .filter((adj) => reach.matrix[adj][targetProjectName])
      .forEach((adj) => {
        visited.push(adj);
        queue.push([adj, [...path]]);
      });
  }

  if (path.length > 1) {
    return path.map((n) => graph.nodes[n]);
  } else {
    return [];
  }
}

export function findConstraintsFor(
  depConstraints: DepConstraint[],
  sourceProject: ProjectGraphNode
) {
  return depConstraints.filter((f) => hasTag(sourceProject, f.sourceTag));
}

export function onlyLoadChildren(
  graph: ProjectGraph,
  sourceProjectName: string,
  targetProjectName: string,
  visited: string[]
) {
  if (visited.indexOf(sourceProjectName) > -1) return false;
  return (
    (graph.dependencies[sourceProjectName] || []).filter((d) => {
      if (d.type !== DependencyType.dynamic) return false;
      if (d.target === targetProjectName) return true;
      return onlyLoadChildren(graph, d.target, targetProjectName, [
        ...visited,
        sourceProjectName,
      ]);
    }).length > 0
  );
}

export function getSourceFilePath(sourceFileName: string, projectPath: string) {
  return path.normalize(sourceFileName).substring(projectPath.length + 1);
}

/**
 * Verifies whether the given node has an architect builder attached
 * @param projectGraph the node to verify
 */
export function hasBuildExecutor(projectGraph: ProjectGraphNode): boolean {
  return (
    // can the architect not be defined? real use case?
    projectGraph.data.targets &&
    projectGraph.data.targets.build &&
    projectGraph.data.targets.build.executor !== ''
  );
}
