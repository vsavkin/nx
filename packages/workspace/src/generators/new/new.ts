import {
  Tree,
  formatFiles,
  updateJson,
  addDependenciesToPackageJson,
  installPackagesTask,
  getWorkspacePath,
  convertNxGenerator,
  names,
  getPackageManagerCommand,
} from '@nrwl/devkit';

import { join } from 'path';
import * as yargsParser from 'yargs-parser';
import { spawn, SpawnOptions } from 'child_process';

import { workspaceGenerator } from '../workspace/workspace';
import { nxVersion } from '../../utils/versions';

export enum Preset {
  Empty = 'empty',
  OSS = 'oss',
  WebComponents = 'web-components',
  Angular = 'angular',
  AngularWithNest = 'angular-nest',
  React = 'react',
  ReactWithExpress = 'react-express',
  NextJs = 'next',
  Nest = 'nest',
}

export interface Schema {
  cli: 'nx' | 'angular';
  directory: string;
  name: string;
  appName: string;
  npmScope?: string;
  skipInstall?: boolean;
  skipGit?: boolean;
  style?: string;
  nxCloud?: boolean;
  preset: Preset;
  commit?: { name: string; email: string; message?: string };
  defaultBase: string;
  linter: 'tslint' | 'eslint';
  packageManager?: string;
}

function generatePreset(host: Tree, opts: Schema) {
  const cliCommand = opts.cli === 'angular' ? 'ng' : 'nx';
  const parsedArgs = yargsParser(process.argv, {
    boolean: ['interactive'],
  });
  const spawnOptions = {
    stdio: [process.stdin, process.stdout, process.stderr],
    shell: true,
    cwd: join(host.root, opts.directory),
  };
  const pmc = getPackageManagerCommand();
  const executable = `${pmc.exec} ${cliCommand}`;
  const args = [
    `g`,
    `@nrwl/workspace:preset`,
    `--name=${opts.appName}`,
    opts.style ? `--style=${opts.style}` : null,
    opts.linter ? `--linter=${opts.linter}` : null,
    opts.npmScope ? `--npmScope=${opts.npmScope}` : `--npmScope=${opts.name}`,
    opts.preset ? `--preset=${opts.preset}` : null,
    `--cli=${cliCommand}`,
    parsedArgs.interactive ? '--interactive=true' : '--interactive=false',
  ].filter((e) => !!e);
  return new Promise((resolve, reject) => {
    spawn(executable, args, spawnOptions).on('close', (code: number) => {
      if (code === 0) {
        resolve();
      } else {
        const message = 'Workspace creation failed, see above.';
        reject(new Error(message));
      }
    });
  });
}

async function initializeGitRepo(
  host: Tree,
  rootDirectory: string,
  options: Schema
) {
  const execute = (args: ReadonlyArray<string>, ignoreErrorStream = false) => {
    const outputStream = 'ignore';
    const errorStream = ignoreErrorStream ? 'ignore' : process.stderr;
    const spawnOptions: SpawnOptions = {
      stdio: [process.stdin, outputStream, errorStream],
      shell: true,
      cwd: join(host.root, rootDirectory),
      env: {
        ...process.env,
        ...(options.commit.name
          ? {
              GIT_AUTHOR_NAME: options.commit.name,
              GIT_COMMITTER_NAME: options.commit.name,
            }
          : {}),
        ...(options.commit.email
          ? {
              GIT_AUTHOR_EMAIL: options.commit.email,
              GIT_COMMITTER_EMAIL: options.commit.email,
            }
          : {}),
      },
    };
    return new Promise((resolve, reject) => {
      spawn('git', args, spawnOptions).on('close', (code) => {
        if (code === 0) {
          resolve();
        } else {
          reject(code);
        }
      });
    });
  };
  const hasCommand = await execute(['--version']).then(
    () => true,
    () => false
  );
  if (!hasCommand) {
    return;
  }
  const insideRepo = await execute(
    ['rev-parse', '--is-inside-work-tree'],
    true
  ).then(
    () => true,
    () => false
  );
  if (insideRepo) {
    console.info(
      `Directory is already under version control. Skipping initialization of git.`
    );
    return;
  }
  await execute(['init']);
  await execute(['add', '.']);
  if (options.commit) {
    const message = options.commit.message || 'initial commit';
    await execute(['commit', `-m "${message}"`]);
  }
  console.info('Successfully initialized git.');
}

export async function newGenerator(host: Tree, options: Schema) {
  if (
    options.skipInstall &&
    options.preset !== 'empty' &&
    options.preset !== 'oss'
  ) {
    throw new Error(`Cannot select a preset when skipInstall is set to true.`);
  }
  if (options.skipInstall && options.nxCloud) {
    throw new Error(`Cannot select nxCloud when skipInstall is set to true.`);
  }

  options = normalizeOptions(options);

  const layout: 'packages' | 'apps-and-libs' =
    options.preset === 'oss' ? 'packages' : 'apps-and-libs';
  const workspaceOpts = {
    ...options,
    layout,
    preset: undefined,
    nxCloud: undefined,
  };
  workspaceGenerator(host, workspaceOpts);

  if (options.cli === 'angular') {
    setDefaultPackageManager(host, options);
  }
  setDefaultLinter(host, options);
  addPresetDependencies(host, options);
  addCloudDependencies(host, options);

  await formatFiles(host);
  host.listChanges().forEach((change) => {
    if (change.type !== 'DELETE') {
      host.rename(change.path, join(options.directory, change.path));
    }
  });
  return async () => {
    installPackagesTask(host, false, options.directory);
    await generatePreset(host, options);
    if (!options.skipGit) {
      await initializeGitRepo(host, options.directory, options);
    }
  };
}

export default newGenerator;
export const newSchematic = convertNxGenerator(newGenerator);

function addCloudDependencies(host: Tree, options: Schema) {
  if (options.nxCloud) {
    return addDependenciesToPackageJson(
      host,
      {},
      { '@nrwl/nx-cloud': 'latest' }
    );
  }
}

const presetDependencies: Omit<
  Record<
    Preset,
    { dependencies: Record<string, string>; dev: Record<string, string> }
  >,
  Preset.Empty | Preset.OSS
> = {
  [Preset.WebComponents]: { dependencies: {}, dev: { '@nrwl/web': nxVersion } },
  [Preset.Angular]: { dependencies: { '@nrwl/angular': nxVersion }, dev: {} },
  [Preset.AngularWithNest]: {
    dependencies: { '@nrwl/angular': nxVersion },
    dev: { '@nrwl/nest': nxVersion },
  },
  [Preset.React]: {
    dependencies: {},
    dev: {
      '@nrwl/react': nxVersion,
    },
  },
  [Preset.ReactWithExpress]: {
    dependencies: {},
    dev: {
      '@nrwl/react': nxVersion,
      '@nrwl/express': nxVersion,
    },
  },
  [Preset.Nest]: {
    dependencies: {},
    dev: {
      '@nrwl/nest': nxVersion,
    },
  },
  [Preset.NextJs]: {
    dependencies: {},
    dev: {
      '@nrwl/next': nxVersion,
    },
  },
};

function addPresetDependencies(host: Tree, options: Schema) {
  if (options.preset === Preset.Empty || options.preset === Preset.OSS) {
    return;
  }
  const { dependencies, dev } = presetDependencies[options.preset];
  return addDependenciesToPackageJson(host, dependencies, dev);
}

function normalizeOptions(options: Schema): Schema {
  options.name = names(options.name).fileName;
  if (!options.directory) {
    options.directory = options.name;
  }

  return options;
}

function setDefaultLinter(host: Tree, { linter, preset }: Schema) {
  // Don't do anything if someone doesn't pick angular
  if (preset !== 'angular' && preset !== 'angular-nest') {
    return;
  }

  switch (linter) {
    case 'eslint': {
      setESLintDefault(host);
      break;
    }
    case 'tslint': {
      setTSLintDefault(host);
      break;
    }
  }
}

/**
 * This sets ESLint as the default for any schematics that default to TSLint
 */
function setESLintDefault(host: Tree) {
  updateJson(host, getWorkspacePath(host), (json) => {
    setDefault(json, '@nrwl/angular', 'application', 'linter', 'eslint');
    setDefault(json, '@nrwl/angular', 'library', 'linter', 'eslint');
    setDefault(
      json,
      '@nrwl/angular',
      'storybook-configuration',
      'linter',
      'eslint'
    );
    return json;
  });
}

/**
 * This sets TSLint as the default for any schematics that default to ESLint
 */
function setTSLintDefault(host: Tree) {
  updateJson(host, getWorkspacePath(host), (json) => {
    setDefault(json, '@nrwl/workspace', 'library', 'linter', 'tslint');
    setDefault(json, '@nrwl/cypress', 'cypress-project', 'linter', 'tslint');
    setDefault(json, '@nrwl/cypress', 'cypress-project', 'linter', 'tslint');
    setDefault(json, '@nrwl/node', 'application', 'linter', 'tslint');
    setDefault(json, '@nrwl/node', 'library', 'linter', 'tslint');
    setDefault(json, '@nrwl/nest', 'application', 'linter', 'tslint');
    setDefault(json, '@nrwl/nest', 'library', 'linter', 'tslint');
    setDefault(json, '@nrwl/express', 'application', 'linter', 'tslint');
    setDefault(json, '@nrwl/express', 'library', 'linter', 'tslint');

    return json;
  });
}

function setDefaultPackageManager(host: Tree, { packageManager }: Schema) {
  if (!packageManager) {
    return;
  }

  updateJson(host, getWorkspacePath(host), (json) => {
    if (!json.cli) {
      json.cli = {};
    }
    json.cli['packageManager'] = packageManager;
    return json;
  });
}

function setDefault(
  json: any,
  collectionName: string,
  generatorName: string,
  key: string,
  value: any
) {
  if (!json.schematics) json.schematics = {};
  if (
    json.schematics[collectionName] &&
    json.schematics[collectionName][generatorName]
  ) {
    json.schematics[collectionName][generatorName][key] = value;
  } else if (json.schematics[`${collectionName}:${generatorName}`]) {
    json.schematics[`${collectionName}:${generatorName}`][key] = value;
  } else {
    json.schematics[collectionName] = json.schematics[collectionName] || {};
    json.schematics[collectionName][generatorName] = { [key]: value };
  }
}
