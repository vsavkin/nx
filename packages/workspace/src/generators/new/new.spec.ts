import { createTree } from '@nrwl/devkit/testing';
import { readJson, Tree } from '@nrwl/devkit';
import { newGenerator, Preset, Schema } from './new';
import { Linter } from '../../utils/lint';

const defaultOptions: Omit<Schema, 'name' | 'directory' | 'appName'> = {
  cli: 'nx',
  preset: Preset.Empty,
  skipInstall: false,
  skipGit: false,
  linter: Linter.EsLint,
  defaultBase: 'master',
};

describe('new', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTree();
  });

  it('should generate an empty workspace.json', async () => {
    await newGenerator(tree, {
      ...defaultOptions,
      name: 'my-workspace',
      directory: 'my-workspace',
      npmScope: 'npmScope',
      appName: 'app',
    });
    expect(readJson(tree, 'my-workspace/workspace.json')).toMatchSnapshot();
  });

  it('should generate an empty nx.json', async () => {
    await newGenerator(tree, {
      ...defaultOptions,
      name: 'my-workspace',
      directory: 'my-workspace',
      npmScope: 'npmScope',
      appName: 'app',
    });
    expect(readJson(tree, 'my-workspace/nx.json')).toMatchSnapshot();
  });

  describe('--preset', () => {
    describe.each([[Preset.Empty], [Preset.Angular], [Preset.React]])(
      '%s',
      (preset) => {
        beforeEach(async () => {
          await newGenerator(tree, {
            ...defaultOptions,
            name: 'my-workspace',
            directory: 'my-workspace',
            npmScope: 'npmScope',
            appName: 'app',
            preset,
          });
        });

        it('should generate necessary npm dependencies', () => {
          expect(readJson(tree, 'my-workspace/package.json')).toMatchSnapshot();
        });
      }
    );
  });
});
