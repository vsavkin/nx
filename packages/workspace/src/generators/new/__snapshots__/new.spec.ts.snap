// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new --preset angular should generate necessary npm dependencies 1`] = `
Object {
  "dependencies": Object {
    "@nrwl/angular": "*",
  },
  "devDependencies": Object {
    "@nrwl/cli": "*",
    "@nrwl/tao": "*",
    "@nrwl/workspace": "*",
    "@types/node": "12.12.38",
    "dotenv": "6.2.0",
    "eslint": "7.10.0",
    "prettier": "2.2.1",
    "ts-node": "~9.1.1",
    "tslint": "~6.1.0",
    "typescript": "~4.0.3",
  },
  "license": "MIT",
  "name": "my-workspace",
  "private": true,
  "scripts": Object {
    "affected": "nx affected",
    "affected:apps": "nx affected:apps",
    "affected:build": "nx affected:build",
    "affected:dep-graph": "nx affected:dep-graph",
    "affected:e2e": "nx affected:e2e",
    "affected:libs": "nx affected:libs",
    "affected:lint": "nx affected:lint",
    "affected:test": "nx affected:test",
    "build": "nx build",
    "dep-graph": "nx dep-graph",
    "e2e": "nx e2e",
    "format": "nx format:write",
    "format:check": "nx format:check",
    "format:write": "nx format:write",
    "help": "nx help",
    "lint": "nx workspace-lint && nx lint",
    "nx": "nx",
    "start": "nx serve",
    "test": "nx test",
    "update": "nx migrate latest",
    "workspace-generator": "nx workspace-generator",
  },
  "version": "0.0.0",
}
`;

exports[`new --preset empty should generate necessary npm dependencies 1`] = `
Object {
  "dependencies": Object {},
  "devDependencies": Object {
    "@nrwl/cli": "*",
    "@nrwl/tao": "*",
    "@nrwl/workspace": "*",
    "@types/node": "12.12.38",
    "dotenv": "6.2.0",
    "eslint": "7.10.0",
    "prettier": "2.2.1",
    "ts-node": "~9.1.1",
    "tslint": "~6.1.0",
    "typescript": "~4.0.3",
  },
  "license": "MIT",
  "name": "my-workspace",
  "private": true,
  "scripts": Object {
    "affected": "nx affected",
    "affected:apps": "nx affected:apps",
    "affected:build": "nx affected:build",
    "affected:dep-graph": "nx affected:dep-graph",
    "affected:e2e": "nx affected:e2e",
    "affected:libs": "nx affected:libs",
    "affected:lint": "nx affected:lint",
    "affected:test": "nx affected:test",
    "build": "nx build",
    "dep-graph": "nx dep-graph",
    "e2e": "nx e2e",
    "format": "nx format:write",
    "format:check": "nx format:check",
    "format:write": "nx format:write",
    "help": "nx help",
    "lint": "nx workspace-lint && nx lint",
    "nx": "nx",
    "start": "nx serve",
    "test": "nx test",
    "update": "nx migrate latest",
    "workspace-generator": "nx workspace-generator",
  },
  "version": "0.0.0",
}
`;

exports[`new --preset react should generate necessary npm dependencies 1`] = `
Object {
  "dependencies": Object {},
  "devDependencies": Object {
    "@nrwl/cli": "*",
    "@nrwl/react": "*",
    "@nrwl/tao": "*",
    "@nrwl/workspace": "*",
    "@types/node": "12.12.38",
    "dotenv": "6.2.0",
    "eslint": "7.10.0",
    "prettier": "2.2.1",
    "ts-node": "~9.1.1",
    "tslint": "~6.1.0",
    "typescript": "~4.0.3",
  },
  "license": "MIT",
  "name": "my-workspace",
  "private": true,
  "scripts": Object {
    "affected": "nx affected",
    "affected:apps": "nx affected:apps",
    "affected:build": "nx affected:build",
    "affected:dep-graph": "nx affected:dep-graph",
    "affected:e2e": "nx affected:e2e",
    "affected:libs": "nx affected:libs",
    "affected:lint": "nx affected:lint",
    "affected:test": "nx affected:test",
    "build": "nx build",
    "dep-graph": "nx dep-graph",
    "e2e": "nx e2e",
    "format": "nx format:write",
    "format:check": "nx format:check",
    "format:write": "nx format:write",
    "help": "nx help",
    "lint": "nx workspace-lint && nx lint",
    "nx": "nx",
    "start": "nx serve",
    "test": "nx test",
    "update": "nx migrate latest",
    "workspace-generator": "nx workspace-generator",
  },
  "version": "0.0.0",
}
`;

exports[`new should generate an empty nx.json 1`] = `
Object {
  "affected": Object {
    "defaultBase": "master",
  },
  "implicitDependencies": Object {
    ".eslintrc.json": "*",
    "nx.json": "*",
    "package.json": Object {
      "dependencies": "*",
      "devDependencies": "*",
    },
    "tsconfig.base.json": "*",
    "tslint.json": "*",
    "workspace.json": "*",
  },
  "npmScope": "npmScope",
  "projects": Object {},
  "tasksRunnerOptions": Object {
    "default": Object {
      "options": Object {
        "cacheableOperations": Array [
          "build",
          "lint",
          "test",
          "e2e",
        ],
      },
      "runner": "@nrwl/workspace/tasks-runners/default",
    },
  },
}
`;

exports[`new should generate an empty workspace.json 1`] = `
Object {
  "cli": Object {
    "defaultCollection": "@nrwl/workspace",
  },
  "projects": Object {},
  "version": 2,
}
`;
